feat: Support building offline OS bundles for Oracle Linux 9.4

manoj-nutanix opened the pull request:
**What problem does this PR solve?**:
Supports building offline OS bundle for Oracle Linux 9.4 


**Which issue(s) does this PR fix?**:
<!-- Add a link to the JIRA issue for both items below
* jql=key in (NCN-102633)
-->
* https://jira.nutanix.com/browse/NCN-102633


**Special notes for your reviewer**:
<!--
Use this to provide any additional information to the reviewers.
This may include:
- Manual testing steps.
- Best way to review the PR.
- Where the author wants the most review attention on.
- etc.
-->

Tested with CI jobs -
1. https://github.com/mesosphere/konvoy-image-builder/actions/runs/11174525665
2. https://github.com/mesosphere/konvoy-image-builder/actions/runs/11175175038

Pending:
Testing cluster creation using offline OS bundle.

**Does this PR introduce a user-facing change?**:
<!--
If yes, add a message in the 'release-note' block below.
If this PR fixes a COPS ticket, include it after the note like: "CLI: Some bug fix. (COPS-xxxx)"
-->
```release-note

```

dkoshkin commented:
The Containerd tar is named https://packages.d2iq.com/dkp/containerd/containerd-1.7.22-d2iq.1-ol-9.4-x86_64.tar.gz, but the e2e test automation assumes its this:
```
Downloading https://packages.d2iq.com/dkp/containerd/containerd-1.7.22-d2iq.1-oracle-9.4-x86_64.tar.gz to artifacts/containerd-1.7.22-d2iq.1-oracle-9.4-x86_64.tar.gz
```

https://github.com/mesosphere/konvoy-image-builder/actions/runs/11105421677/job/30851548106#step:10:1693 

You'll want another change to replace `oracle` with `ol` in https://github.com/mesosphere/konvoy-image-builder/blob/4f299425c6ffee8f2b1f9ecdcce890d1da33becd/magefile.go#L454

manoj-nutanix commented:
> The Containerd tar is named https://packages.d2iq.com/dkp/containerd/containerd-1.7.22-d2iq.1-ol-9.4-x86_64.tar.gz, but the e2e test automation assumes its this:
> 
> ```
> Downloading https://packages.d2iq.com/dkp/containerd/containerd-1.7.22-d2iq.1-oracle-9.4-x86_64.tar.gz to artifacts/containerd-1.7.22-d2iq.1-oracle-9.4-x86_64.tar.gz
> ```
> 
> https://github.com/mesosphere/konvoy-image-builder/actions/runs/11105421677/job/30851548106#step:10:1693
> 
> You'll want another change to replace `oracle` with `ol` in
> 
> https://github.com/mesosphere/konvoy-image-builder/blob/4f299425c6ffee8f2b1f9ecdcce890d1da33becd/magefile.go#L454

Thanks, will make this change.. will go through e2e test to check the steps and packages it requires.

dkoshkin commented:
(Updated the PR title because it gets used by automation for the release)

manoj-nutanix commented on bundles/redhat8.6/packages.txt.gotmpl:
```diff
@@ -11,7 +11,6 @@ sysstat
 yum-utils
 yum-plugin-versionlock
 NetworkManager
-yum-utils
```
duplicate values for `yum-utils`  .. similar change in other files

manoj-nutanix commented on cmd/konvoy-image-wrapper/cmd/create-package-bundle.go:
```diff
@@ -47,6 +47,10 @@ var osToConfig = map[string]OSConfig{
 	},
+	"oracle-9.4": {
+		configDir:      "bundles/oracle9.4",
+		containerImage: "docker.io/library/oraclelinux:9",
```
the minor version container images have not been released yet on any registry([ref](https://hub.docker.com/_/oraclelinux/tags?page_size=&ordering=&name=9.)) so using this image which currently points to 9.4 only -

```
[root@8b53e13be278 generated]# cat /etc/oracle-release
Oracle Linux Server release 9.4

[root@8b53e13be278 generated]# cat /etc/redhat-release 
Red Hat Enterprise Linux release 9.4 (Plow)

[root@8b53e13be278 generated]# cat /etc/os-release 
NAME="Oracle Linux Server"
VERSION="9.4"
ID="ol"
ID_LIKE="fedora"
VARIANT="Server"
VARIANT_ID="server"
VERSION_ID="9.4"
PLATFORM_ID="platform:el9"
PRETTY_NAME="Oracle Linux Server 9.4"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:oracle:linux:9:4:server"
HOME_URL="https://linux.oracle.com/"
BUG_REPORT_URL="https://github.com/oracle/oracle-linux"

ORACLE_BUGZILLA_PRODUCT="Oracle Linux 9"
ORACLE_BUGZILLA_PRODUCT_VERSION=9.4
ORACLE_SUPPORT_PRODUCT="Oracle Linux"
ORACLE_SUPPORT_PRODUCT_VERSION=9.4
```

manoj-nutanix replied:
we can update the image tag once 9.4 version is released on their official registry/repo -
1. https://container-registry.oracle.com/ords/f?p=113:4:10433988665656:::4:P4_REPOSITORY,AI_REPOSITORY,AI_REPOSITORY_NAME,P4_REPOSITORY_NAME,P4_EULA_ID,P4_BUSINESS_AREA_ID:1,1,Oracle%20Linux,Oracle%20Linux,1,0&cs=3GLgmu4zPDtIjZjF1cXhfpBMVzRl0Sefp1HfIY1Aw88xKXzFyVBlZp58aVGnejQ6Yin9zyBXV3AgvO8P-JhWwgw
2. https://github.com/oracle/container-images/pkgs/container/oraclelinux

Alternatively we can peg image tag to specific SHA but I'm skeptical about reliability of SHA tags to be generally available all time -
<img width="1512" alt="Screenshot 2024-10-03 at 4 16 07 PM" src="https://github.com/user-attachments/assets/89cc11b1-03dc-470c-bceb-c7a7c8c8ae8b">

manoj-nutanix commented:
I'd squashed commits, which mistakenly included [PR](https://github.com/mesosphere/konvoy-image-builder/pull/1190) changes as well in single commit.. will rebase/remove GHA changes before merging.

manoj-nutanix commented on .pre-commit-config.yaml:
```diff
@@ -56,7 +56,7 @@ repos:
         stages: [commit]
   - repo: https://github.com/golangci/golangci-lint
-    rev: v1.43.0
+    rev: v1.61.0
```
it was failing earlier so upgraded -
```
golangci-lint............................................................Failed
- hook id: golangci-lint
- exit code: 2

panic: load embedded ruleguard rules: rules/rules.go:13: can't load fmt

goroutine 1 [running]:
github.com/go-critic/go-critic/checkers.init.9()
        /Users/manoj.surudwad/.cache/pre-commit/repofcru5w2o/golangenv-default/pkg/mod/github.com/go-critic/go-critic@v0.6.1/checkers/checkers.go:58 +0x424
```

dkoshkin replied:
Can you pull this out in a separate PR please, we'll need to backport this

manoj-nutanix replied:
oh, will do and rebase too.

manoj-nutanix replied:
done in 1cfdfac2ae78a972651a758ff298519dc6fff60a

dkoshkin commented:
Looks great, I just merged https://github.com/mesosphere/konvoy-image-builder/pull/1190, can you please rebase on `main`.
You can also remove the non-offline tests from the GHA files (like we have for the other OSs), I don't think we need both in there.

manoj-nutanix commented:
> Looks great, I just merged #1190, can you please rebase on `main`. You can also remove the non-offline tests from the GHA files (like we have for the other OSs), I don't think we need both in there.

rebased and removed the non-offline tests for oracle-9.4 .. I suppose there would some e2e test covering both offline and non-offline modes of OS image as there're few ansible tasks which are specific to modes so confirming.

Also, I've not ran test for oracle-9.4 `offline-fips` mode.. shall I do it or not needed?

supershal commented on .github/workflows/release-vsphere-template.yaml:
```diff
@@ -28,9 +28,9 @@ jobs:
             buildConfig: "offline"
           - os: "oracle 9.4"
-            buildConfig: "basic"
+            buildConfig: "offline"
```
we should add `offline` and `offline-fips` in addition to `basic` and `fips`. we need to test for all 4 buildconfigs

manoj-nutanix replied:
thanks, I have done it as per @dkoshkin suggestion as we're doing the same for other OS images(only offline* modes are tested) - https://github.com/mesosphere/konvoy-image-builder/pull/1192#issuecomment-2392004946 .. lmk if we want to run for all 4 buildconfigs.

dkoshkin replied:
We don't have all 4 for the other OSs, but we could start here

manoj-nutanix replied:
done f521da9876a38f6d214b3a77066bb597ff6894d3

supershal replied:
Thanks. I would prefer atleast one flavor of `basic` tests. Sometimes we do miss failure with `networked` tests.

manoj-nutanix commented on .github/workflows/vsphere-e2e.yaml:
```diff
@@ -32,9 +32,9 @@ jobs:
         - os: "rocky 9.1"
           buildConfig: "offline"
         - os: "oracle 9.4"
-          buildConfig: "basic"
```
https://github.com/mesosphere/konvoy-image-builder/pull/1192#discussion_r1786705437

manoj-nutanix commented on ansible/roles/containerd/tasks/redhat.yaml:
```diff
@@ -41,6 +40,7 @@
     - container-selinux
   when:
     - ansible_distribution == 'OracleLinux'
+    - not offline_mode_enabled
```
done in f59ff616972e0c48f8882de17859961af2766744

supershal commented:
Running vsphere e2e tests. Good to merge if those are successful for Oracle 9.4: https://github.com/mesosphere/konvoy-image-builder/actions/runs/11186167572/

dkoshkin commented on ansible/roles/containerd/tasks/redhat.yaml:
```diff
@@ -26,12 +26,13 @@
     state: present
     update_cache: true
+    enablerepo: "{{ 'offline' if offline_mode_enabled else '' }}"
+    disablerepo: "{{ '*' if offline_mode_enabled else '' }}"
```
Good catch.
